rosbag2_compression/src/rosbag2_compression/zstd_decompressor.cpp

```diff
@@ -0,0 +1,242 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <cstdio>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "rcpputils/filesystem_helper.hpp"
+
+#include "rosbag2_compression/zstd_decompressor.hpp"
+
+#include "rosbag2_storage/filesystem_helper.hpp"
+
+#include "logging.hpp"
+
+namespace
+{
+
+// String constant used to identify ZstdDecompressor.
+constexpr const char kDecompressionIdentifier[] = "zstd";
+
+/**
+ * Open a file using the C API.
+ * This function calls OS-specific implementation of fopen.
+ *
+ * \param uri is the path to the file
+ * \param read_mode is the read mode string accepted by fopen.
+ * \return the FILE pointer or nullptr if the file was not opened.
+ */
+FILE * open_file(const std::string & uri, const std::string & read_mode)
+{
+  FILE * fp{nullptr};
+#ifdef _WIN32
+  fopen_s(&fp, uri.c_str(), read_mode.c_str());
+#else
+  fp = std::fopen(uri.c_str(), read_mode.c_str());
+#endif
+  return fp;
+}
+
+/**
+ * Read a file from the supplied uri into a vector.
+ *
+ * \param uri is the path to the file.
+ * \return the contents of the buffer as a vector.
+ */
+std::vector<uint8_t> get_input_buffer(const std::string & uri)
+{
+  // Get the file size
+  const auto compressed_buffer_length = rosbag2_storage::FilesystemHelper::get_file_size(uri);
+
+  // Read in buffer, handling accordingly
+  const auto file_pointer = open_file(uri.c_str(), "rb");
+  if (file_pointer == nullptr) {
+    std::stringstream errmsg;
+    errmsg << "Failed to open file: \"" << uri << "\" for binary reading!";
+
+    throw std::runtime_error{errmsg.str()};
+  }
+
+  // Initialize compress_buffer with size = compressed_buffer_length.
+  // Uniform initialization cannot be used here since it will choose
+  // the initializer list constructor instead.
+  std::vector<uint8_t> compressed_buffer(compressed_buffer_length);
+
+  const auto nRead = fread(
+    compressed_buffer.data(), sizeof(decltype(compressed_buffer)::value_type),
+    compressed_buffer_length, file_pointer);
+
+  if (nRead != compressed_buffer_length) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM("Bytes read !(" <<
+      nRead << ") != compressed_buffer_length (" << compressed_buffer_length <<
+      ")!");
+    // An error indicator is set by fread, so the following check will throw.
```

Comment:
This comment actually refers to line 90, right?

Reply:
Yeah, the error log message on L84 exists to provide further information on an error from `fread`.

The comment on L87 is indicating that the code should throw at this state and that the check on L90 does the actual throw since `fread` may have set the error indicator for other reasons besides `nRead != compressed_buffer_length`.